### PR TITLE
feat: Add compilation in custom location(folder)

### PR DESCRIPTION
**Pull request by taco-paco**
@satyambnsal [Notice](https://github.com/NethermindEth/zksync-remix-plugin/blob/2ff673e4e712a38654b0cc4c1357818795bf89a2/api/src/handlers/compile.rs#L168) that the decision of if we shall display contracts from non-standard location shall be taken on Front level. Here I list all of the .sol contracts even outside `/contracts` folder

```python
if !single_file:
    display only from ./contracts
else: 
    display all artifacts
```

### CODE REVIEW

File: api/src/handlers/verify.rs [hunk 1]
```diff
@@ -58,9 +58,27 @@ pub async fn get_verify_result(process_id: String, engine: &State<WorkerEngine>)
         _ => String::from("Result not available"),
     })
 }
```
Comment (varex83) [inline api/src/handlers/verify.rs:60]: add a blank line here)

**Review by varex83 — APPROVED** [review]
LGTM

File: api/src/handlers/verify.rs (hunk 1 above)
Comment (taco-paco) [reply api/src/handlers/verify.rs:60]: Makes sense) Done

File: plugin/src/stores/remixClient.ts [hunk 2]
```diff
@@ -8,7 +8,12 @@ const WORKSPACE_ROOT = '.workspaces'
 const remixClientAtom = atom({} as unknown as RemixClient)
 const noFileSelectedAtom = atom(false)
 const isValidSolidityAtom = atom(false)
-const currentFilenameAtom = atom('')
+const currentFilepathAtom = atom('')
+const currentFilenameAtom = atom((get) => {
```
Comment (taco-paco) [inline plugin/src/stores/remixClient.ts:12]: @satyambnsal There's no really a need in an atom for this. When it's used we can explicitly call `getFileNameFromPath` on `currentFilepathAtom`. Right now when reading `Compilation.txs` there's impression that those 2 can have different states.

File: plugin/src/features/Compilation/index.tsx [hunk 3]
```diff
@@ -32,49 +34,60 @@ export const Compilation = ({ setAccordian }: CompilationProps) => {
...
     await remixClient.editor.clearAnnotations()
     try {
-      const workspaceContents = {
-        config: {
-          version: solidityVersion,
-          user_libraries: []
-        },
-        contracts: [] as Array<{ file_name: string; file_content: string; is_contract: boolean }>
+      const allContractFiles = await getAllContractFiles(remixClient, currentWorkspacePath)
+      console.log('All contract files', allContractFiles)
```
Comment (taco-paco) [inline plugin/src/features/Compilation/index.tsx:75]: @satyambnsal This is debug log I guess. Print all the files. Shall be removed in prod

File: plugin/src/features/Compilation/index.tsx [hunk 4]
```diff
@@ -32,49 +34,60 @@ export const Compilation = ({ setAccordian }: CompilationProps) => {
...
+      workspaceContents.contracts = allContractFiles
+
+      if (type === 'SINGLE_FILE') {
+        const targetPath = getContractTargetPath(allContractFiles, currentFilepath)
```
Comment (taco-paco) [inline plugin/src/features/Compilation/index.tsx:79]: @satyambnsal Why do we iterate over all contracts here? We already know `contractFilepath`. We shall just use this logic

```
 const parts = file_name.split('/')
      if (parts.length === 1) {
        return './'
      } else {
        parts.pop()
        return './' + parts.join('/')
      }
```

File: plugin/src/utils/remix_storage.ts [hunk 5]
```diff
@@ -19,3 +19,25 @@ export const getAllContractFiles = async (remixClient: RemixClient, path: string
...
+        parts.pop()
+        return './' + parts.join('/')
+      }
+    }
+  }
+
+  return './'
+}
+
+export const findFilesNotInContracts = (allContracts: ContractFile[]) => {
```
Comment (taco-paco) [inline plugin/src/utils/remix_storage.ts:39]: @satyambnsal do we want to find all the files or only contracts?

File: plugin/src/stores/remixClient.ts (hunk 2 above)
Comment (satyambnsal) [reply plugin/src/stores/remixClient.ts:12]: Its a derived read only atom where the atom value gets derived from `currentFilepathAtom`.

File: plugin/src/utils/remix_storage.ts (hunk 5 above)
Comment (satyambnsal) [reply plugin/src/utils/remix_storage.ts:39]: Only contract files.

File: plugin/src/features/Compilation/index.tsx (hunk 4 above)
Comment (satyambnsal) [reply plugin/src/features/Compilation/index.tsx:79]: Good suggestion. Thank you!

File: plugin/src/stores/remixClient.ts (hunk 2 above)
Comment (taco-paco) [reply plugin/src/stores/remixClient.ts:12]: Yeah, I saw that, but I don't think there's an atom needed for that since the only atom we need is `filePath`

**Pull request merged**